metaflow/client/core.py

```diff
@@ -1123,6 +1124,202 @@ def _iter_filter(self, x):
         # exclude private data artifacts
         return x.id[0] != "_"
 
+    def _get_task_for_queried_step(self, flow_id, run_id, query_step):
+        """
+        Returns a Task object corresponding to the queried step.
+        If the queried step has several tasks, the first task is returned.
+        """
+        # Find any task corresponding to the queried step
+        step = Step(f"{flow_id}/{run_id}/{query_step}", _namespace_check=False)
+        task = next(iter(step.tasks()), None)
+        if task:
+            return task
+        raise MetaflowNotFound(f"No task found for the queried step {query_step}")
+
+    def _get_metadata_query_vals(
+        self,
+        flow_id: str,
+        run_id: str,
+        cur_foreach_stack_len: int,
+        steps: List[str],
+        query_type: str,
+    ):
+        """
+        Returns the field name and field value to be used for querying metadata of successor or ancestor tasks.
+
+        Parameters
+        ----------
+        flow_id : str
+            Flow ID of the task
+        run_id : str
+            Run ID of the task
+        cur_foreach_stack_len : int
+            Length of the foreach stack of the current task
+        steps : List[str]
+            List of step names whose tasks will be returned. For static joins, and static splits, we can have
+            ancestors and successors across multiple steps.
+        query_type : str
+            Type of query. Can be 'ancestor' or 'successor'.
+        """
+        # For each task, we also log additional metadata fields such as foreach-indices and foreach-indices-truncated
+        # which help us in querying ancestor and successor tasks.
+        #       `foreach-indices`: contains the indices of the foreach stack at the time of task execution.
+        #       `foreach-indices-truncated`: contains the indices of the foreach stack at the time of task execution but
+        #       truncated by 1
+        # For example, a task thats nested 3 levels deep in a foreach stack may have the following values:
+        # foreach-indices = [0, 1, 2]
+        # foreach-indices-truncated = [0, 1]
+
+        if len(steps) > 1:
+            # This is a static join or a static split. There will be no change in foreach stack length
+            query_foreach_stack_len = cur_foreach_stack_len
+        else:
+            # For linear steps, or foreach splits and joins, ancestor and successor tasks will all belong to
+            # the same step.
+            query_task = self._get_task_for_queried_step(flow_id, run_id, steps[0])
+            query_foreach_stack_len = len(
+                query_task.metadata_dict.get("foreach-indices", [])
+            )
+
+        if query_foreach_stack_len == cur_foreach_stack_len:
+            # The successor or ancestor tasks belong to the same foreach stack level
+            field_name = "foreach-indices"
+            field_value = self.metadata_dict.get(field_name)
```

Comment:
We don't currently cache metadata_dict so either we could fix that or cache it here to avoid making multiple calls to the metadata service and then sorts. It would need to be cached across _get_related_tasks and this function.